src/transformers/models/flava/modeling_flava.py

```diff
@@ -0,0 +1,2273 @@
+                output is returned when `bool_masked_pos` has some of the patches masked.
+        mmm_text_logits (`torch.FloatTensor` of shape `(batch_size, text_seq_length, text_vocab_size)` or of shape `(`(total_masked_seq_length, text_vocab_size)`), *optional*, returned when `pixel_values` and `input_ids_masked` are present):
+                The logits for MMM text multimodal loss. The flattened output is returned when `input_ids_masked` has
+                some of the tokens masked.
+        contrastive_logits_per_image (`torch.FloatTensor` of shape `(image_batch_size, text_batch_size)`):
+            The scaled dot product scores between `image_embeddings` and `text_embeddings` but passed through FLAVA's
+            `image_projection` and `text_projection` layers respectively. This represents the image-text similarity
+            scores. This is calculated on unmasked images and texts.
+        contrastive_logits_per_text (`torch.FloatTensor` of shape `(text_batch_size, image_batch_size)`):
+            The scaled dot product scores between `text_embeddings` and `image_embeddings` but passed through FLAVA's
+            `text_projection` and `image_projection` layers respectively. This is calculated on unmasked images and
+            texts.
+    """
+
+    loss: Optional[torch.FloatTensor] = None
+    loss_info: FlavaLosses = None
+    image_embeddings: Optional[torch.FloatTensor] = None
+    image_output: Optional[BaseModelOutputWithPooling] = None
+    text_embeddings: Optional[torch.FloatTensor] = None
+    text_output: Optional[BaseModelOutputWithPooling] = None
+    multimodal_embeddings: Optional[torch.FloatTensor] = None
+    multimodal_output: Optional[BaseModelOutputWithPooling] = None
+    image_masked_embeddings: Optional[torch.FloatTensor] = None
+    image_masked_output: Optional[BaseModelOutputWithPooling] = None
+    text_masked_embeddings: Optional[torch.FloatTensor] = None
+    text_masked_output: Optional[BaseModelOutputWithPooling] = None
+    multimodal_masked_embeddings: Optional[torch.FloatTensor] = None
+    multimodal_masked_output: Optional[BaseModelOutputWithPooling] = None
+    mim_logits: Optional[torch.FloatTensor] = None
+    mlm_logits: Optional[torch.FloatTensor] = None
+    itm_logits: Optional[torch.FloatTensor] = None
+    contrastive_logits_per_image: Optional[torch.FloatTensor] = None
+    contrastive_logits_per_text: Optional[torch.FloatTensor] = None
+    mmm_image_logits: Optional[torch.FloatTensor] = None
+    mmm_text_logits: Optional[torch.FloatTensor] = None
+
+    def to_tuple(self) -> Tuple[Any]:
+        transformer_outputs = [
+            "text_output",
+            "image_output",
+            "multimodal_output",
+            "text_masked_output",
+            "image_masked_output",
+            "multimodal_masked_output",
+        ]
+        return tuple(self[k] if k not in transformer_outputs else getattr(self, k).to_tuple() for k in self.keys())
+
+
+def add_masked_text(
```

Comment:
This is and `add_whole_word_masked_text` are not linked to the model per se, so they should go somewhere else (probably in an example script showing how to use FLAVA). It's as if the word masking code was in the BERT file (which it's not).

Reply:
That's only relevant for pretraining no?

Reply:
So, I am somewhat conflicted about these. I think Flava is incomplete without these and I am also using them in the tests to verify that pretraining loss calculation logic is working as expected. These are actually similar in spirit to the processing functions in modeling_wav2vec2 https://github.com/huggingface/transformers/blob/259eeb6dabf0457bd935de0729e292b2c728a990/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L194. I would ideally like to keep them if possible and I can move them to `processing_flava` so as not to pollute modeling code if that helps. What do you think?

Reply:
Note that while `_compute_mask_indices` is also used for pretraining in Wav2Vec2, it's main purpose is to apply SpecAugment during fine-tuning which is similar in spirit to dropout. Also no tokenizer logic is used in `_compute_mask_indices`, it's therefore not dependent on any settings of the tokenizer, but just the model.

Reply:
The logic for mlm is inside a data collator and people usually these days want to prove the pretrain model for their pretraining objectives as well (I want to probe it actually as well), so that’s why I believe having this as a separate function would make sense (even if only in processing_flava file). Another question, how would I do the integration test for users without these function, hardcode masked indices?

Reply:
I have removed these for now. I will look into some other alternative to provide them in easier way later.